### PR TITLE
Permissions refactor

### LINKED ISSUE #199: Populate DMS with permissions

### LINKED ISSUE #302: New permissions system
TODO list:
- [x] Create models for `Permission` and `Group`
- [x] Add `Permission` codename template validation
- [x] Create API to add/remove permissions from `User`
- [x] Create API to add/remove `Group`s from `User`
- [x] Create API to create `Group`s, add/remove permissions from `Group`
- [x] Create API to transfer group permissions to user, removing group from user
- [x] Create API to remove `User` permissions in bulk
- [x] Implement `SELF` scope in `Teacher`
- [x] Implement `MILGROUP` scope in `Teacher`
- [x] Implement `MILFACULTY` scope in `Teacher`
- [x] Create inline registering for `Permission`s
- [x] Generalize scope queryset filtering
- [x] Scopes for `POST`
- [x] Scopes for `PUT`
- [x] Scopes for `PATCH`
- [x] Scopes for `DELETE`
- [x] Add perms & scopes to `Absence`
- [x] Add perms & scopes to `Achievement`
- [x] Add perms & scopes to `Encouragement`
- [x] Add perms & scopes to `Lesson`
- [x] Add perms & scopes to `Mark`
- [x] Add perms & scopes to `Punishment`
- [x] Add perms & scopes to `Reference Book`
- [x] Add perms & scopes to `Student`
- [x] Add perms & scopes to `Uniform`
- [x] Add perms & scopes to `Subject`
- [x] Rewrite `Teacher` perms
- [x] Add permissions to modify `Permission`s and `Group`s
- [x] Add permissions and scopes to `DMS` views?

### PR DESCRIPTION
**Pull request by gakhromov**
Resolves #302 and closes #199.

### CODE REVIEW

File: back-end/src/auth/models.py [hunk 1]
```diff
@@ -28,6 +32,49 @@ def create_superuser(self, email, password, **extra_fields):
...
+        MILFACULTY = 10, "milfaculty"
+        MILGROUP = 20, "milgroup"
+        SELF = 30, "self"
+
+    scope = models.IntegerField(choices=Scopes.choices)
+    viewset = models.CharField(max_length=100)
+    method = models.CharField(max_length=100)
+
+    name = models.CharField(max_length=255)
+    codename = models.CharField(max_length=100, unique=True)
```
Comment (gakhromov) [inline back-end/src/auth/models.py:48]: Should make this a computed field

File: back-end/src/auth/populate/permissions.py [hunk 2]
```diff
@@ -1,33 +1,42 @@
...
-        {
-            "codename": view_name + "_get",
-            "name": view_name_rus + get_str,
-        },
-        {
-            "codename": view_name + "_post",
-            "name": view_name_rus + post_str,
-        },
-        {
-            "codename": view_name + "_patch",
```
Comment (gakhromov) [inline back-end/src/auth/populate/permissions.py:20]: Remove codename from here and below

File: back-end/src/auth/populate/permissions.py [hunk 3]
```diff
@@ -1,33 +1,42 @@
...
+            permissions.append({
+                "viewset":
+                    view_name,
+                "method":
+                    method,
+                "scope":
+                    scopes_str[scope][0],
+                "codename":
+                    "_".join([view_name, method, scope]),
+                "name":
```
Comment (gakhromov) [inline back-end/src/auth/populate/permissions.py:31]: Use `"."` separation, as some names may contain underscores

**Review by gakhromov — COMMENTED** [review]
Rebase + tmp remove migration

**Comment by gakhromov** [comment]
Add a check for the same permissions with different scopes

File: back-end/src/auth/management/commands/register_permissions.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                            scopes_str[scope]
+                        ]),
+                })
+
+        for val in permissions:
+            Permission.objects.get_or_create(
+                viewset=val["viewset"],
+                method=val["method"],
+                scope=val["scope"],
+                name=val["name"],
```
Comment (TmLev) [inline back-end/src/auth/management/commands/register_permissions.py:48]: `**val`

File: back-end/src/auth/models.py [hunk 5]
```diff
@@ -60,5 +110,53 @@ class User(AbstractUser):
...
+
+    def has_general_perm(self, viewset, method):
+        perms = self._filter_permissions(viewset, method)
+        return len(perms) > 0
+
+    def get_perm_scope(self, viewset, method):
+        perms = self._filter_permissions(viewset, method)
+        if len(perms) > 0:
+            return min(perms, key=lambda perm: perm["scope"])["scope"]
+        return None
```
Comment (TmLev) [inline back-end/src/auth/models.py:159]:
```python3
return min([p["scope"] for p in perms]) if perms else None
```

File: back-end/src/auth/urls.py [hunk 6]
```diff
@@ -1,13 +1,49 @@
...
     TokenObtainPairExtendedView,
     TokenRefreshExtendedView,
     ChangePasswordAPIView,
 )
 
+routers = DefaultRouter()
+# Using routers.register does not register DELETE and POST
+# actions properly, so I register them by hand
+routers.register("user-control", UserControlViewSet)
+routers.register("group", GroupViewSet)
```
Comment (TmLev) [inline back-end/src/auth/urls.py:17]: Make it singular (`router`).

File: back-end/src/auth/urls.py [hunk 7]
```diff
@@ -1,13 +1,49 @@
...
+    UserControlViewSet,
+    GroupViewSet,
     TokenObtainPairExtendedView,
     TokenRefreshExtendedView,
     ChangePasswordAPIView,
 )
 
+routers = DefaultRouter()
+# Using routers.register does not register DELETE and POST
+# actions properly, so I register them by hand
```
Comment (TmLev) [inline back-end/src/auth/urls.py:15]: image

File: back-end/src/auth/views.py [hunk 8]
```diff
@@ -15,28 +17,285 @@
...
+    @action(
+        detail=True,
+        methods=["delete"],
+        url_path="permissions",
+    )
+    # pylint: disable=invalid-name, unused-argument
+    def delete_permissions(self, request: Request, pk=None) -> Response:
+        """Delete user permissions."""
+        query_params = PermissionRequestSerializer(data=request.query_params)
+        if not query_params.is_valid():
```
Comment (TmLev) [inline back-end/src/auth/views.py:92]: `.is_valid(raise_exception=True)`

File: back-end/src/auth/views.py [hunk 9]
```diff
@@ -15,28 +17,285 @@
...
+        detail=True,
+        methods=["post"],
+        url_path="permissions",
+    )
+    # pylint: disable=invalid-name, unused-argument
+    def add_permissions(self, request: Request, pk=None) -> Response:
+        """Add user permissions."""
+        data = PermissionRequestSerializer(data=request.data)
+        if not data.is_valid():
+            return Response(data.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (TmLev) [inline back-end/src/auth/views.py:68]: Same as query_params.

File: back-end/src/auth/views.py [hunk 10]
```diff
@@ -15,28 +17,285 @@
...
+    )
+    # pylint: disable=invalid-name, unused-argument
+    def add_groups(self, request: Request, pk=None) -> Response:
+        """Add user to groups."""
+        data = GroupShortSerializer(data=request.data)
+        if data.is_valid():
+            # if such group exists, it violates the unique constraint
+            # and, thus, is invalid. That's what we need to check.
+            # If data is valid, such group does not exist -> bad request
+            return Response(data.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (TmLev) [inline back-end/src/auth/views.py:127]: If `data` is valid, there will be no `data.errors` (?).

File: back-end/src/auth/views.py [hunk 11]
```diff
@@ -15,28 +17,285 @@
...
+    queryset = Group.objects.all()
+    serializer_class = GroupSerializer
+    permission_classes = [PermissionPermission]
+
+    @extend_schema(request=GroupModifySerializer)
+    def create(self, request: Request) -> Response:
+        # validate group name
+        data = GroupModifySerializer(data=request.data)
+        if not data.is_valid():
+            return Response(data.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (TmLev) [inline back-end/src/auth/views.py:214]: `raise_exception`

File: back-end/src/auth/views.py [hunk 12]
```diff
@@ -15,28 +17,285 @@
...
+        # validate permissions
+        perms = [{
+            "codename": codename
+        } for codename in data.data["permissions"]]
+        validation = PermissionRequestSerializer(data=perms, many=True)
+        if not validation.is_valid():
+            return Response(validation.errors, status.HTTP_400_BAD_REQUEST)
```
Comment (TmLev) [inline back-end/src/auth/views.py:221]: `raise_exception`

File: back-end/src/auth/views.py [hunk 13]
```diff
@@ -15,28 +17,285 @@
...
+        detail=True,
+        methods=["post"],
+        url_path="permissions",
+    )
+    # pylint: disable=invalid-name, unused-argument
+    def add_permissions(self, request: Request, pk=None) -> Response:
+        """Add group permissions."""
+        data = PermissionRequestSerializer(data=request.data)
+        if not data.is_valid():
+            return Response(data.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (TmLev) [inline back-end/src/auth/views.py:249]: `raise_exception`

File: back-end/src/auth/views.py [hunk 14]
```diff
@@ -15,28 +17,285 @@
...
+        methods=["delete"],
+        url_path="permissions",
+    )
+    # pylint: disable=invalid-name, unused-argument
+    def delete_permissions(self, request: Request, pk=None) -> Response:
+        """Delete group permissions."""
+        query_params = PermissionRequestSerializer(data=request.query_params)
+        if not query_params.is_valid():
+            return Response(query_params.errors,
+                            status=status.HTTP_400_BAD_REQUEST)
```
Comment (TmLev) [inline back-end/src/auth/views.py:275]: `raise_exception`

File: back-end/src/lms/functions.py [hunk 15]
```diff
@@ -1,5 +1,79 @@
...
+    Returns empty string and None if user is not
+    a teacher or a student.
+
+    :param request: REST/HTTP request
+    :return: user_type (str), user (User object)
+    """
+    # check if user is a teacher or a student
+    user = Teacher.objects.filter(user=request.user)
+    user_type = 'teacher'
+    if user.count() == 0:
```
Comment (TmLev) [inline back-end/src/lms/functions.py:24]: ```if user.exists():```

File: back-end/src/lms/functions.py [hunk 16]
```diff
@@ -1,5 +1,79 @@
...
+        # check if user is a student
+        user = Student.objects.filter(user=request.user)
+        user_type = 'student'
+        if user.count() == 0:
+            return '', None
+    return user_type, user.first()
```
Comment (TmLev) [inline back-end/src/lms/functions.py:30]:
```python3
if (user := Teacher.objects.filter(user=request.user)).exists():
    return "teacher", user

if (user := Student.objects.filter(user=request.user)).exists():
    return "student", user

return "", None
```

File: back-end/src/lms/mixins.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+        if user_type == "teacher":
+            return self.queryset.filter(
+                student__milgroup__milfaculty=user.milfaculty)
+        return self.queryset.none()
+
+    def allow_scope_milfaculty_on_create(self, data, user_type, user):
+        student = Student.objects.filter(id=data["student"])
+        if student.count() == 0:
+            return False
+        if user_type == "student":
```
Comment (TmLev) [inline back-end/src/lms/mixins.py:145]: `if not student.exists():`

**Review by TmLev — APPROVED** [review]
LGTM. Prepare to merge.

**Pull request merged**